pkg/serviceaccounts/controllers/docker_registry_service.go

```diff
@@ -29,188 +33,339 @@ type DockerRegistryServiceControllerOptions struct {
 				return e.client.Services(options.RegistryNamespace).Watch(opts)
 			},
 		},
-		&api.Service{},
+		&kapi.Service{},
 		options.Resync,
 		framework.ResourceEventHandlerFuncs{
-			AddFunc:    e.serviceAdded,
-			UpdateFunc: e.serviceUpdated,
-			DeleteFunc: e.serviceDeleted,
+			AddFunc: func(obj interface{}) {
+				e.enqueueRegistryLocationQueue()
+			},
+			UpdateFunc: func(old, cur interface{}) {
+				e.enqueueRegistryLocationQueue()
+			},
+			DeleteFunc: func(obj interface{}) {
+				e.enqueueRegistryLocationQueue()
+			},
 		},
 	)
-	e.registryServiceName = options.RegistryServiceName
-	e.dockercfgController = options.DockercfgController
-	e.defaultDockerURL = options.DefaultDockerURL
+	e.servicesSynced = e.serviceController.HasSynced
+	e.syncRegistryLocationHandler = e.syncRegistryLocationChange
+
+	dockercfgOptions := kapi.ListOptions{FieldSelector: fields.SelectorFromSet(map[string]string{kapi.SecretTypeField: string(kapi.SecretTypeDockercfg)})}
+	e.secretCache, e.secretController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(opts kapi.ListOptions) (runtime.Object, error) {
+				return e.client.Secrets(kapi.NamespaceAll).List(dockercfgOptions)
+			},
+			WatchFunc: func(opts kapi.ListOptions) (watch.Interface, error) {
+				return e.client.Secrets(kapi.NamespaceAll).Watch(dockercfgOptions)
+			},
+		},
+		&kapi.Secret{},
+		options.Resync,
+		framework.ResourceEventHandlerFuncs{},
+	)
+	e.secretsSynced = e.secretController.HasSynced
+	e.syncSecretHandler = e.syncSecretUpdate
 
 	return e
 }
 
 // DockerRegistryServiceController manages ServiceToken secrets for Service objects
 type DockerRegistryServiceController struct {
-	stopChan chan struct{}
-
 	client client.Interface
 
-	registryServiceName string
-	defaultDockerURL    string
+	serviceName      string
+	serviceNamespace string
 
 	dockercfgController *DockercfgController
 
-	serviceController *framework.Controller
+	serviceController           *framework.Controller
+	serviceCache                cache.Store
+	servicesSynced              func() bool
+	syncRegistryLocationHandler func(key string) error
+
+	secretController  *framework.Controller
+	secretCache       cache.Store
+	secretsSynced     func() bool
+	syncSecretHandler func(key string) error
+
+	registryURLs          sets.String
+	registryURLLock       sync.RWMutex
+	registryLocationQueue workqueue.RateLimitingInterface
+	secretsToUpdate       workqueue.RateLimitingInterface
+
+	dockerURLsIntialized chan struct{}
 }
 
 // Runs controller loops and returns immediately
-func (e *DockerRegistryServiceController) Run() {
-	if e.stopChan == nil {
-		e.stopChan = make(chan struct{})
-		go e.serviceController.Run(e.stopChan)
+func (e *DockerRegistryServiceController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	go e.serviceController.Run(stopCh)
+	go e.secretController.Run(stopCh)
+	go e.waitForDockerURLs(stopCh)
```

Comment:
> Shouldn't we wait for this before starting any work?

No need.  The event is guaranteed to come *after* the cache has been updated to have our entry.  You usually only have to sync when you're feeding from two source.